middleware/stack_values.go

```diff
@@ -0,0 +1,101 @@
+package middleware
+
+import (
+	"context"
+	"reflect"
+	"strings"
+)
+
+// WithStackValue adds a key value pair to the context that are intended to be
+// scoped to a stack. Use ClearStackValues to get a new context with all stack
+// values cleared.
+func WithStackValue(ctx context.Context, key, value interface{}) context.Context {
+	md, _ := ctx.Value(stackValuesKey{}).(*stackValues)
+
+	md = withStackValue(md, key, value)
+	return context.WithValue(ctx, stackValuesKey{}, md)
+}
+
+// ClearStackValues returns a context without any stack values.
+func ClearStackValues(ctx context.Context) context.Context {
+	return context.WithValue(ctx, stackValuesKey{}, nil)
+}
+
+// GetStackValues returns the value pointed to by the key within the stack
+// values, if it is present.
+func GetStackValue(ctx context.Context, key interface{}) interface{} {
+	v := ctx.Value(stackValuesKey{})
+	if v == nil {
+		return nil
+	}
+
+	md := v.(*stackValues)
```

Comment:
I think we actually need to check that md itself is also not nil.

Reply:
As a small example on the different between a nil empty interface and a interface containing a pointer to something that is nil:

```package main

type Foo struct {}

func (f *Foo) DoThing() {
	if f == nil {
		panic("Foo is nil")
	}
}

func main() {
	var x interface{}

	if x == nil {
		println("x is nil")
	}

	x = (*Foo)(nil)

	if x != nil {
		println("x is not nil")
	}

	f := x.(*Foo)
	f.DoThing()
}
```

This prints
```
x is nil
x is not nil
panic: Foo is nil
```

Reply:
Simplified this to the following which solves this issue.

```go
	md, _ := ctx.Value(stackValuesKey{}).(*stackValues)
	if md == nil {
		return nil
	}
```